### PR TITLE
Beautiful log message capturing :star: :heart_eyes_cat:

### LINKED ISSUE #28101: Fix wonkiness with `mt/with-log-level`
`mt/with-log-level` should probably recursively update child loggers e.g. if we have a `metabase` logger and a `metabase.internal` logger, then setting `mt/with-log-level` for `metabase` should change the levels for BOTH. This seems like the obvious behavior. The fact that it **doesn't** work this way seems like a gotcha

### LINKED ISSUE #39335: Address TODOs in `metabase.test.util.log`
https://github.com/metabase/metabase/blob/master/test/metabase/test/util/log.clj#L238
https://github.com/metabase/metabase/blob/master/test/metabase/test/util/log.cljs#L17

### PR DESCRIPTION
**Pull request by camsaul**
New log capturing macro replacement for the old version of `with-log-messages-for-level` and `with-log-level` macros in tests. 

It has the following benefits over the old version:

- Thread safe, safe in parallel tests
- Can use it multiple times to capture multiple different levels of logs that conflict with one another for whatever reason you would want to do this
- Does not permanently alter the loggers or anything weird
- Eliminates the gotcha in #28101 -- recursively affects 'child' namespaces
- Works exactly the same in both Clj and Cljs 
- Results are returned in a much less horrible format, better for writing tests
- Nicer syntax, logs are returned by a function that you bind, rather than as the output of the entire macro form
- This also means you can call that function multiple times. Neat
- Doesn't touch stupid Log4j2 stuff so some random upstream change in Log4j2 isn't going to break all our tests again randomly
- Eliminates spaminess in tests. Manipulating the log level currently causes things to get logged in test output even tho tests are supposed to be QUIET
- Works in ClojureScript tests

Example from the tests:

```clj
(log.capture/with-log-messages-for-level [util-debug-messages ["metabase.util" :debug]
                                          test-trace-messages [metabase.util.log.capture-test :trace]]
  (is (= []
         (util-debug-messages)))
  (is (= []
         (test-trace-messages)))
  (testing "trace message should only show up in :trace"
    (log/tracef "a %s" "picture")
    (is (= []
           (util-debug-messages)))
    (is (= [{:namespace 'metabase.util.log.capture-test
             :level     :trace
             :e         nil
             :message   "a picture"}]
           (test-trace-messages))))
  (testing "debug message should show up in :trace and :debug"
    (log/debugf "a %s" "bug")
    (is (= [{:namespace 'metabase.util.log.capture-test
             :level     :debug
             :e         nil
             :message   "a bug"}]
           (util-debug-messages)))
    (is (= [{:namespace 'metabase.util.log.capture-test
             :level     :trace
             :e         nil
             :message   "a picture"}
            {:namespace 'metabase.util.log.capture-test
             :level     :debug
             :e         nil
             :message   "a bug"}]
           (test-trace-messages)))))
```

The macroexpansion of something like

```clj
(log/trace "a picture")
```

becomes

```clj
(do
  (metabase.util.log.capture/capture-logp
   "metabase.util.log.capture-test"
   :trace
   "a picture")
  ;; clojure.tools.logging stuff
  ...)
```

becomes

```clj
(do
  (when-let [f# (metabase.util.log.capture/*capture-logs-fn*
                 "metabase.util.log.capture-test"
                 5)]
      (metabase.util.log.capture/capture-logp! f# "a picture"))
  ...)
```

Basic idea is we have a dynamic variable called `*capture-logs-fn*` with the signature

```clj
(f namespace-str level-int)
```

and if the logs should be captured at that level, it returns a function with the signature

```clj
(f e message)
```

that you should call with the logged exception (if any) and logged message to capture the message.

Then the actual implementation can basically compose `*capture-logs-fn*` so you can have multiple functions capturing
logs.

The impl can store the logs in an atom or whatever that you can get later.

Resolves #28101
Resolves #39335

### CODE REVIEW

File: src/metabase/util/log.clj [hunk 1]
```diff
@@ -78,17 +86,23 @@
   You shouldn't have to use this directly; prefer the level-specific macros like [[info]]."
   {:arglists '([level message & more] [level throwable message & more])}
   [level x & more]
-  (macros/case
-    :cljs (glogi-logp (str *ns*) level x more)
-    :clj  (tools-logp *ns*       level x more)))
+  `(do
```
Comment (dpsutton) [inline src/metabase/util/log.clj:89]: I don't think you need the do here, you could just have this top level since the body of a defn has an implicit do

Comment (camsaul) [reply src/metabase/util/log.clj:89]: This is a `defn` that's used for macroexpansion output

File: src/metabase/util/log.cljs [hunk 2]
```diff
@@ -8,6 +8,12 @@
   (:require-macros
    [metabase.util.log]))
 
+;; if capture stuff is available, make sure it's loaded, because the macros in the Clojure version of this namespace
+;; refer to it
+(try
+  (require 'metabase.util.log.capture)
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/util/log.cljs]: Yup, turned out this didn't work

**Review by bshepherdson — COMMENTED** [review]
😭  beautiful!

I hope we can find a good way to fix the CLJS lack of dynamic `require`.

File: .clj-kondo/config.edn [hunk 3]
```diff
@@ -616,13 +616,11 @@
   clojure.core.logic/defne                                                             clj-kondo.lint-as/def-catch-all
   clojure.test.check.clojure-test/defspec                                              clojure.test/deftest
   clojurewerkz.quartzite.jobs/defjob                                                   clojure.core/defn
-  honeysql.helpers/defhelper                                                           clj-kondo.lint-as/def-catch-all
```
Comment (camsaul) [inline .clj-kondo/config.edn:619]: removed since we don't use Honey SQL 1 anymore

File: .clj-kondo/config.edn [hunk 4]
```diff
@@ -616,13 +616,11 @@
...
-  honeysql.util/defalias                                                               clojure.core/def
-  metabase-enterprise.serialization.test-util/with-random-dump-dir                     clojure.core/let
   metabase-enterprise.serialization.test-util/with-dbs                                 clojure.core/fn
+  metabase-enterprise.serialization.test-util/with-random-dump-dir                     clojure.core/let
```
Comment (camsaul) [inline .clj-kondo/config.edn:620]: sorted the rest of these

File: .clj-kondo/config.edn [hunk 5]
```diff
@@ -688,20 +686,21 @@
...
    clojure.test/deftest                                                                                                      hooks.clojure.test/deftest
    clojure.test/is                                                                                                           hooks.clojure.test/is
    clojure.test/use-fixtures                                                                                                 hooks.clojure.test/use-fixtures
-   metabase-enterprise.cache.config-test/with-temp-persist-models                                                            hooks.common/with-seven-bindings
    metabase-enterprise.advanced-permissions.models.permissions.application-permissions-test/with-new-group-and-current-graph hooks.common/with-two-top-level-bindings
    metabase-enterprise.audit-app.pages-test/with-temp-objects                                                                hooks.common/with-one-binding
+   metabase-enterprise.cache.config-test/with-temp-persist-models                                                            hooks.common/with-seven-bindings
```
Comment (camsaul) [inline .clj-kondo/config.edn:698]: sorted these

**Comment by piranha** [comment]
Just in case, there is [some stuff](https://github.com/metabase/metabase/blob/serdes-extract-nested/src/metabase/logger.clj#L127) I did for serialization API, maybe worth taking a look. Now that I'm looking at it I'm almost sure it's not thread safe though 🤦‍♂️

File: src/metabase/util/log/capture.cljc [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+;;;      (when-let [f# (metabase.util.log.capture/*capture-logs-fn*
+;;;                     "metabase.util.log.capture-test"
+;;;                     5)]
+;;;          (metabase.util.log.capture/capture-logp! f# "a picture"))
+;;;      ...)
+;;;
+;;; VERY VERY IMPORTANT! [[*capture-logs-fn*]] only returns a function if logs for that namespace AND level are to be
+;;; captured! Only then is [[capture-logp!]] or [[capture-logf!]], which ultimately call the returned function `f#`,
+;;; called with the args to the logging macro -- and only then do they get evaluated! It's super important not to
+;;; evaluate the log args unless we actually want to capture them! They might be very expensive!
```
Comment (piranha) [inline src/metabase/util/log/capture.cljc:173]: Hmmm, why expensive? Is this about `print-str` and overall string generation or more?

File: test/metabase/api/database_test.clj [hunk 7]
```diff
@@ -2094,16 +2094,17 @@
                      :unaggregated-query-row-limit (symbol "nil #_\"key is not present.\"")}
                     (settings)))))))))
 
-(deftest log-an-error-if-contains-undefined-setting-test
+(deftest ^:paralell log-an-error-if-contains-undefined-setting-test
```
Comment (piranha) [inline test/metabase/api/database_test.clj:2097]: `:parallel`

**Review by sansolovyov — APPROVED** [review]
Like the syntax! I'm approving (and it'll get merged), but a few `paralell`s need fixing anyway.

I can't decide if this approach makes sense to use in serialization API instead of log4j2 utilities I made, but it writes to a file, so maybe not. 😁

**Pull request merged**